File: trump_game/code7.py
import random

class Card:
      def __init__(self, value, suit, intValue) :
            self.value = value
            self.suit = suit
            self.intValue = intValue

      def getCardString(self):
            return self.suit + self.value + "(" + str(self.intValue) + ")"

class Deck:
      def __init__(self):
            self.deck = self.generateDeck()

      def printDeck(self):
            print("Displaying cards...")
            for i in self.deck:
                  print(i.getCardString())

      def shuffleDeck(self):
            deckSize = len(self.deck)
            for i in range(0, deckSize):
                  j = random.randint(0,i)
                  temp = self.deck[i]
                  self.deck[i] = self.deck[j]
                  self.deck[j] = temp

      def draw(self):
            return self.deck.pop()

      @staticmethod
      def generateDeck():
            newDeck = []
            values = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]
            suits = ["♣", "♦", "♥", "♠"]

            for suit in suits:
                  for i, value in enumerate(values):
                        newDeck.append(Card(value, suit, i))

            return newDeck

class HelperFunction:
      @staticmethod
      def maxInArrayIndex(points):
            maxInt = [0,points[0]]
            for index, point in enumerate(points[1:]):
                  if int(maxInt[1]) < point:
                        maxInt = [index + 1, point]
            return maxInt[0]


# ディーラークラスの続きです。tabelのgameModeによって、勝利条件を変える関数を作成しましょう。
class Dealer:

      @staticmethod
      def initialTable(amoutOfPeople ,gameMode):
            table = {
                  "players": [],
                  "gameMode": gameMode,
                  "deck": Deck()
            }

            table["deck"].shuffleDeck()

            for person in range(0, amoutOfPeople):
                  personCards = []
                  for cards in range(0, Dealer.initialCards(gameMode)):
                        personCards.append(table["deck"].draw())
                  table["players"].append(personCards)

            return table


      @staticmethod
      def initialCards(gameMode):
            if gameMode == "21":
                  return 2
            if gameMode == "porker":
                  return 5

      # ここから記入してください
      # テーブルの状態を与え、ゲームの種類に応じた勝利条件にを返します・
      @staticmethod
      def selectWinnerConditions(table):
            if table["gameMode"] == "21":
                  return Dealer.checkWinner21(table)
            if table["gameMode"] == "porker":
                  return "no game"

      @staticmethod
      def checkWinner21(table):
            playerPoints = []
            cachePoints = {}
            for cards in table["players"]:
                  point = Dealer.score21Individual(cards)
                  playerPoints.append(point)
                  if point in cachePoints:
                        cachePoints[point] += 1
                  else:
                        cachePoints[point] = 1

            maxIndex = HelperFunction.maxInArrayIndex(playerPoints)

            if cachePoints[playerPoints[maxIndex]] == 0:
                  return "no winner"
            elif cachePoints[playerPoints[maxIndex]] > 1:
                  return "draw"
            else:
                  return "winner is " + str(maxIndex + 1) + "player"

      @staticmethod
      def score21Individual(cards):
            value = 0
            for card in cards:
                  value += card.intValue + 1

            return value if 21 >= value >= 1 else 0
      
      @staticmethod
      def printTable(table):
            print("This table: ")
            print("gameMode: " + table["gameMode"])
            print("deck: ")
            table["deck"].printDeck()
            for i, player in enumerate(table["players"]):
                  print(str(i + 1) + "player's cards: ")
                  for cards in player:
                        print(cards.getCardString())

# テーブルを生成します
table1 = Dealer.initialTable(3, "21")
Dealer.printTable(table1)
# ゲームの種類によって勝利条件を選び、勝者を返します
print(Dealer.selectWinnerConditions(table1))

File: trump_game_js/code7.js
class Card {
      constructor(suit, value, intValue) {
            this.suit = suit;
            this.value = value;
            this.intValue = intValue
      }

      getCardString() {
            return this.suit + this.value + "(" + this.intValue + ")";
      }
}

class Deck {
      constructor(deck) {
            this.deck = deck;
      }

      shuffleDeck() {
            for (let i = 0; i < this.deck.length; i++) {
                  let j = Math.floor(Math.random() * (i + 1));
                  let temp = this.deck[i];
                  this.deck[i] = this.deck[j];
                  this.deck[j] = temp;
            }
      }

      printDeck() {
            console.log("Displaying cards...")
            for (let i = 0; i < this.deck.length; i++) {
                  console.log(this.deck[i].getCardString());
            }
      }

      draw() {
            return this.deck.pop()
      }

      static generateDeck() {
            let newDeck = [];
            const suits = ["♣", "♦", "♥", "♠"];
            const values = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]

            for (let i = 0; i < suits.length; i++) {
                  for (let j = 0; j < values.length; j++) {
                        newDeck.push(new Card(suits[i], values[j], j));
                  }
            }

            return newDeck;
      }
}

class HelperFunction {
      static helperMaxInt(points) {
            let maxInt = [0,0];
            for (let i = 0; i < points.length; i++) {
                  if (maxInt[1] < points[i]) { 
                        maxInt[1] = points[i];
                        maxInt[0] = i;
                  }
            }
            return maxInt;
      }
}

// ディーラークラスの続きです。tabelのgameModeによって、勝利条件を変える関数を作成しましょう。

class Dealer {

      static startGame(amountOfPlayers, gameMode) {
            let table = {
                  "players":[],
                  "gameMode": gameMode,
                  "deck": new Deck(Deck.generateDeck())
            }

            table["deck"].shuffleDeck();

            for (let i = 0; i < amountOfPlayers; i++) {
                  let playerCard = [];
                  for (let j = 0; j < Dealer.initialCards(gameMode); j++) {
                        playerCard.push(table["deck"].draw());
                  }
                  table["players"].push(playerCard);
            }

            return table;
      }

      static initialCards(gameMode) {
            if (gameMode == "21") { return 2;}
            if (gameMode == "porker") { return 5;}
      }

      static score21Individual(cards) {
            let value = 0;
            for (let i = 0; i < cards.length; i++) {
                  value += cards[i].intValue + 1;
            }
            if (value > 21) { value = 0; }
            return value;
      }

      // ここから記入してください
      // テーブルの状態を与え、ゲームの種類に応じた勝利条件にを返します・
      static checkWinner(table) {
            if (table["gameMode"] == "21") {
                  return Dealer.checkWinner21(table);
            } else {
                  return "no game"
            }
      }

      static checkWinner21(table) {
            let playerStats = [];
            let cachedScores = [];
            for (let i = 0; i < table["players"].length; i++) {
                  let player = table["players"][i];
                  let point = Dealer.score21Individual(player);
                  if (cachedScores[point] >= 1) {
                        cachedScores[point] += 1;
                  } else {
                        cachedScores[point] = 1;
                  }
                  playerStats.push(point);
            }

            let maxInt = HelperFunction.helperMaxInt(playerStats);

            if (maxInt[1] == 0) {
                  return "No winners..";
            } else if (cache[maxInt[1]] > 1) {
                  return "It is a draw ";
            } else {
                  return "player " + (maxInt[0] + 1) + " is the winner";
            }

      }
}

// 勝利条件の記述があるゲームの場合
let table1 = Dealer.startGame(3, "21");
console.log(Dealer.checkWinner(table1));
// 勝利条件の記述がないゲームの場合
let table2 = Dealer.startGame(3, "porker");
console.log(Dealer.checkWinner(table2));